elfpy/data/contract_interface.py

```diff
@@ -420,3 +435,87 @@ def get_hyperdrive_config(hyperdrive_contract: Contract) -> PoolConfig:
     out_config["termLength"] = term_length
 
     return PoolConfig(**out_config)
+
+
+def get_wallet_info(
+    hyperdrive_contract: Contract,
+    base_contract: Contract,
+    block_number: BlockNumber,
+    transactions: list[Transaction],
+) -> list[WalletInfo]:
+    """Retrieves wallet information at a given block given a transaction
+    Transactions are needed here to get
+    (1) the wallet address of a transaction, and
+    (2) the token id of the transaction
+
+    """
+
+    # pylint: disable=too-many-locals
+
+    out_wallet_info = []
+    for transaction in transactions:
+        wallet_addr = transaction.event_operator
+        token_id = transaction.event_id
+        token_prefix = transaction.event_prefix
+        token_maturity_time = transaction.event_maturity_time
+
+        if wallet_addr is None:
+            continue
+
+        num_base_token_fp = None
+        for _ in range(RETRY_COUNT):
+            try:
+                num_base_token_fp = base_contract.functions.balanceOf(wallet_addr).call(block_identifier=block_number)
+                break
+            except ValueError:
+                logging.warning("Error in getting base token balance, retrying")
+                time.sleep(1)
+                continue
+
+        num_base_token = convert_fixedpoint(num_base_token_fp)
+        if (num_base_token is not None) and (wallet_addr is not None):
+            out_wallet_info.append(
+                WalletInfo(
+                    blockNumber=block_number,
+                    walletAddress=wallet_addr,
+                    baseTokenType="BASE",
+                    tokenType="BASE",
+                    tokenValue=num_base_token,
+                )
+            )
+
+        # Handle cases where these fields don't exist
+        if (token_id is not None) and (token_prefix is not None):
+            base_token_type = hyperdrive_assets.AssetIdPrefix(token_prefix).name
+            if (token_maturity_time is not None) and (token_maturity_time > 0):
+                token_type = base_token_type + "-" + str(token_maturity_time)
+                maturity_time = token_maturity_time
+            else:
+                token_type = base_token_type
+                maturity_time = None
+
+            num_custom_token_fp = None
+            for _ in range(RETRY_COUNT):
+                try:
+                    num_custom_token_fp = hyperdrive_contract.functions.balanceOf(int(token_id), wallet_addr).call(
+                        block_identifier=block_number
+                    )
+                except ValueError:
+                    logging.warning("Error in getting custom token balance, retrying")
+                    time.sleep(1)
+                    continue
+            num_custom_token = convert_fixedpoint(num_custom_token_fp)
```

Comment:
what does `convert_fixedpoint` do?  is `num_custom_token_fp` a FixedPoint or floating point?  can't the FixedPoint class convert to float already?

Reply:
convert_fixedpoint is the way I'm doing scaled_values -> unscaled_value using the FixedPoint class. It also handles the case when the input is None.